src/main/java/picard/illumina/BarcodeExtractor.java

```diff
@@ -0,0 +1,308 @@
+        final byte[][] perfectScores = new byte[readStructure.sampleBarcodes.length()][];
+        int index = 0;
+        for (final ReadDescriptor d : readStructure.descriptors) {
+            if (d.type == ReadType.Barcode) {
+                perfectScores[index] = new byte[d.length];
+                Arrays.fill(perfectScores[index], (byte) 60);
+                noMatchBarcode[index++] = StringUtil.repeatCharNTimes('N', d.length);
+            }
+        }
+
+        this.noMatch = new BarcodeMetric(null, null, IlluminaUtil.barcodeSeqsToString(noMatchBarcode), noMatchBarcode);
+
+        Set<BarcodeExtractor.ByteString> barcodesBytes = new HashSet<>(barcodeToMetrics.size());
+        for (final BarcodeMetric metric : barcodeToMetrics.values()) {
+            barcodesBytes.add(new BarcodeExtractor.ByteString(metric.barcodeBytes));
+            metrics.put(metric.BARCODE_WITHOUT_DELIMITER, metric);
+        }
+        this.barcodesBytes = barcodesBytes;
+
+        // Prepopulate the lookup map with all perfect barcodes
+        for(BarcodeMetric metric : barcodeToMetrics.values()){
+            BarcodeExtractor.BarcodeMatch match = calculateBarcodeMatch(metric.barcodeBytes,
+                    perfectScores, true);
+            barcodeLookupMap.put(new BarcodeExtractor.ByteString(metric.barcodeBytes), match);
+        }
+
+        // Prepopulate all no call barcode match
+        BarcodeExtractor.BarcodeMatch noCallMatch = calculateBarcodeMatch(noMatchMetric.barcodeBytes,
+                perfectScores, true);
+
+        barcodeLookupMap.put(new BarcodeExtractor.ByteString(noMatchMetric.barcodeBytes), noCallMatch);
+    }
+
+    public Map<String, BarcodeMetric> getMetrics() {
+        return this.metrics;
+    }
+
+    public BarcodeMetric getNoMatchMetric() {
+        return this.noMatch;
+    }
+
+    /**
+     * Find the best barcode match for the given read sequence, and accumulate metrics
+     *
+     * NOTE: the returned BarcodeMatch object will contain mismatches mismatchesToSecondBest values that may be
+     * inaccurate as long as the conclusion match/no-match isn't affected. for example, mismatches and mismatchesToSecondBest
+     * may be smaller than their true value if mismatches is truly larger than maxMismatches.
+     * Also, mismatchesToSecondBest might be smaller than its true value if its true value is greater than
+     * mismatches + minMismatchDelta. This is due to an optimization which allows the distance calculation to stop once
+     * the conclusion (Match or no-Match) can be reached.
+     *
+     * @param readSubsequences portion of read containing barcode
+     * @return perfect barcode string, if there was a match within tolerance, or null if not.
+     */
+    BarcodeMatch findBestBarcode(final byte[][] readSubsequences,
+                                 final byte[][] qualityScores,
+                                 final boolean isInlineMatching) {
```

Comment:
From context I can't figure out what `isInlineMatching` does - could you add an `@param` for it please?

Reply:
Ok, I figured it out (duh @tfenne ) but still think some parameter documentation on what it is and what effects it has would be good.